### DISCUSSION TITLE
Using Flipper successfully in system tests

### DISCUSSION THREAD

**Discussion by jhubert** [jnunemaker/flipper#695] [Q&A]
Is there a best practice for using Flipper in system tests? Our system tests run in a different process, so the memory adapter from [the documentation](https://www.flippercloud.io/docs/testing#starting-fresh) doesn't work. The test `Flipper.instance` is different from the app `Flipper.instance` that's being tested. I don't think we're doing anything that crazy... just using cuprite as a driver on top of headless chrome.

Is there a best practice we should be following for this? Or a way that other people have done it? I read through [the closed issue on this topic](https://github.com/jnunemaker/flipper/issues/615) but there doesn't seem to be any resolution.

**Comment by jhubert** [comment]
This is what we're using now, based on [the closed issue](https://github.com/jnunemaker/flipper/issues/615), and it seems to work but there are concerns about thread safety.

In the flipper.rb initializer:
```ruby

if Rails.env.test?
  shared_test_adapter = Flipper::Adapters::Memory.new

  Flipper.configure do |config|
    config.adapter do
      shared_test_adapter
    end
  end
end
```

In our test helper:
```ruby
  def after_teardown
    super

    Flipper.instance.adapter.get_all.clear
  end
```

**Comment by jnunemaker** [comment]
@jhubert if you are using a separate process and need to share state across the processes, you can use the [pstore adapter](https://github.com/jnunemaker/flipper/blob/main/lib/flipper/adapters/pstore.rb) (assuming they are on the same machine, here is an example of [test setup](https://github.com/jnunemaker/flipper/blob/e98695c64458d5cbea313c8c0caab810162d2f5c/test/adapters/pstore_test.rb#L7-L12)). You can also use the active record adapter for system tests. 

The primary reason we recommend the memory adapter for tests is speed and cleanliness of state between tests, but the active record adapter should retain this as well assuming you are using a transaction around or database cleaner around each test. 

Does any of that help or can we drop more specifics? Just let us know.

**Reply by jhubert** [reply]
Thanks @jnunemaker. That makes sense. 

Any suggestions for how to configure this properly? It sounds like the ideal would be memory adapter in most tests but pstore in system tests, or is the perf difference not that big?

**Reply by jnunemaker** [reply]
Very hard to say. Best way to find out would be to change adapter and see if the test run duration goes up a bunch or not. Should be an easy check.
